File: internal/builder/builder.go
package builder

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"docksmith/internal/cache"
	"docksmith/internal/layers"
	"docksmith/internal/models"
	"docksmith/internal/parser"
	"docksmith/internal/runtime"
	"docksmith/internal/state"
)

// BuildImage is the main entry point for `docksmith build`.
func BuildImage(contextDir, name, tag string, noCache bool) (*models.ImageManifest, error) {
	docksmithfilePath := filepath.Join(contextDir, "Docksmithfile")
	instructions, err := parser.ParseDocksmithfile(docksmithfilePath)
	if err != nil {
		return nil, err
	}

	totalSteps := len(instructions)
	var layerEntries []models.LayerEntry
	envDict := make(map[string]string)
	workdir := ""
	var cmd []string
	prevDigest := ""
	cacheBusted := false
	buildStart := time.Now().UTC()
	var originalCreated string

	cm := cache.NewManager(noCache)

	fmt.Println()

	for stepIdx, instr := range instructions {
		label := fmt.Sprintf("Step %d/%d : %s %s", stepIdx+1, totalSteps, instr.Type, instr.Args)

		switch instr.Type {
		case "FROM":
			fmt.Println(label)
			baseName, baseTag := parser.ParseFromArgs(instr.Args)
			baseManifest, err := state.LoadManifest(baseName, baseTag)
			if err != nil {
				return nil, err
			}
			if baseManifest == nil {
				return nil, fmt.Errorf("[BUILD ERROR] FROM: image %q not found.\n  Run setup-base first to import base images.", instr.Args)
			}
			layerEntries = append([]models.LayerEntry(nil), baseManifest.Layers...)
			prevDigest = baseManifest.Digest
			for _, envStr := range baseManifest.Config.Env {
				if idx := strings.Index(envStr, "="); idx >= 0 {
					envDict[envStr[:idx]] = envStr[idx+1:]
				}
			}
			workdir = baseManifest.Config.WorkingDir
			cmd = baseManifest.Config.Cmd

			if existing, _ := state.LoadManifest(name, tag); existing != nil {
				originalCreated = existing.Created
			}

		case "WORKDIR":
			fmt.Println(label)
			workdir = instr.Args

		case "ENV":
			fmt.Println(label)
			k, v, err := parser.ParseEnvArgs(instr.Args)
			if err != nil {
				return nil, err
			}
			envDict[k] = v

		case "CMD":
			fmt.Println(label)
			cmd, err = parser.ParseCmdArgs(instr.Args)
			if err != nil {
				return nil, err
			}

		case "COPY":
			src, dest, err := parser.ParseCopyArgs(instr.Args)
			if err != nil {
				return nil, err
			}
			instructionText := "COPY " + instr.Args
			envSer := cache.SerializeEnv(envDict)
			copyHashes, err := layers.HashCopySources(src, contextDir)
			if err != nil {
				return nil, err
			}

			cachedDigest := ""
			if !cacheBusted {
				cachedDigest = cm.Lookup(prevDigest, instructionText, workdir, envSer, copyHashes)
			}

			if cachedDigest != "" {
				fmt.Printf("%s [CACHE HIT]\n", label)
				sz, _ := layers.GetLayerSize(cachedDigest)
				layerEntries = append(layerEntries, models.LayerEntry{
					Digest:    cachedDigest,
					Size:      sz,
					CreatedBy: instructionText,
				})
				prevDigest = cachedDigest
			} else {
				cacheBusted = true
				t0 := time.Now()
				digest, err := executeCopy(src, dest, contextDir, layerEntries, workdir)
				if err != nil {
					return nil, err
				}
				elapsed := time.Since(t0)
				fmt.Printf("%s [CACHE MISS] %.2fs\n", label, elapsed.Seconds())

				cm.Store(prevDigest, instructionText, workdir, envSer, copyHashes, digest)
				sz, _ := layers.GetLayerSize(digest)
				layerEntries = append(layerEntries, models.LayerEntry{
					Digest:    digest,
					Size:      sz,
					CreatedBy: instructionText,
				})
				prevDigest = digest
			}

		case "RUN":
			instructionText := "RUN " + instr.Args
			envSer := cache.SerializeEnv(envDict)

			cachedDigest := ""
			if !cacheBusted {
				cachedDigest = cm.Lookup(prevDigest, instructionText, workdir, envSer, nil)
			}

			if cachedDigest != "" {
				fmt.Printf("%s [CACHE HIT]\n", label)
				sz, _ := layers.GetLayerSize(cachedDigest)
				layerEntries = append(layerEntries, models.LayerEntry{
					Digest:    cachedDigest,
					Size:      sz,
					CreatedBy: instructionText,
				})
				prevDigest = cachedDigest
			} else {
				cacheBusted = true
				t0 := time.Now()
				digest, err := executeRun(instr.Args, layerEntries, envDict, workdir)
				if err != nil {
					return nil, err
				}
				elapsed := time.Since(t0)
				fmt.Printf("%s [CACHE MISS] %.2fs\n", label, elapsed.Seconds())

				cm.Store(prevDigest, instructionText, workdir, envSer, nil, digest)
				sz, _ := layers.GetLayerSize(digest)
				layerEntries = append(layerEntries, models.LayerEntry{
					Digest:    digest,
					Size:      sz,
					CreatedBy: instructionText,
				})
				prevDigest = digest
			}
		}
	}

	// Build env list
	envList := make([]string, 0, len(envDict))
	for k, v := range envDict {
		envList = append(envList, k+"="+v)
	}

	allHits := !cacheBusted
	created := buildStart.Format(time.RFC3339Nano)
	if allHits && originalCreated != "" {
		created = originalCreated
	}

	manifest := models.ImageManifest{
		Name:    name,
		Tag:     tag,
		Digest:  "",
		Created: created,
		Config: models.ImageConfig{
			Env:        envList,
			Cmd:        cmd,
			WorkingDir: workdir,
		},
		Layers: layerEntries,
	}

	saved, err := state.SaveManifest(manifest)
	if err != nil {
		return nil, err
	}

	totalTime := time.Since(buildStart)
	shortDigest := saved.Digest
	if len(shortDigest) > 19 {
		shortDigest = shortDigest[:19]
	}
	fmt.Printf("\nSuccessfully built %s %s:%s (%.2fs)\n\n", shortDigest, name, tag, totalTime.Seconds())
	return &saved, nil
}

func assembleRootfs(layerEntries []models.LayerEntry, targetDir string) error {
	for _, layer := range layerEntries {
		if err := layers.ExtractLayer(layer.Digest, targetDir); err != nil {
			return err
		}
	}
	return nil
}

func executeCopy(srcPattern, dest, contextDir string, currentLayers []models.LayerEntry, workdir string) (string, error) {
	rootfs, err := os.MkdirTemp("", "docksmith_copy_")
	if err != nil {
		return "", err
	}
	defer os.RemoveAll(rootfs)

	if err := assembleRootfs(currentLayers, rootfs); err != nil {
		return "", err
	}

	if workdir != "" {
		os.MkdirAll(filepath.Join(rootfs, strings.TrimPrefix(workdir, "/")), 0755)
	}

	destAbs := filepath.Join(rootfs, strings.TrimPrefix(dest, "/"))
	os.MkdirAll(destAbs, 0755)

	// Resolve source files
	var matched []string
	if srcPattern == "." {
		err := filepath.Walk(contextDir, func(p string, info os.FileInfo, err error) error {
			if err != nil {
				return err
			}
			if !info.IsDir() {
				matched = append(matched, p)
			}
			return nil
		})
		if err != nil {
			return "", err
		}
	} else {
		pattern := filepath.Join(contextDir, srcPattern)
		matched, err = filepath.Glob(pattern)
		if err != nil {
			return "", err
		}
	}

	if len(matched) == 0 {
		return "", fmt.Errorf("[COPY ERROR] no files matched %q in %q", srcPattern, contextDir)
	}

	for _, srcPath := range matched {
		info, err := os.Stat(srcPath)
		if err != nil {
			continue
		}
		rel, _ := filepath.Rel(contextDir, srcPath)
		if info.IsDir() {
			copyDir(srcPath, filepath.Join(destAbs, filepath.Base(srcPath)))
		} else {
			dstDir := filepath.Join(destAbs, filepath.Dir(rel))
			os.MkdirAll(dstDir, 0755)
			copyFile(srcPath, filepath.Join(dstDir, filepath.Base(srcPath)))
		}
	}

	allPaths, err := layers.CollectAllPaths(destAbs)
	if err != nil {
		return "", err
	}
	tarBytes, err := layers.CreateDeltaTar(rootfs, allPaths)
	if err != nil {
		return "", err
	}
	return layers.StoreLayer(tarBytes)
}

func executeRun(command string, currentLayers []models.LayerEntry, envDict map[string]string, workdir string) (string, error) {
	rootfs, err := os.MkdirTemp("", "docksmith_run_")
	if err != nil {
		return "", err
	}
	defer os.RemoveAll(rootfs)

	if err := assembleRootfs(currentLayers, rootfs); err != nil {
		return "", err
	}

	effectiveWorkdir := workdir
	if effectiveWorkdir == "" {
		effectiveWorkdir = "/"
	}
	os.MkdirAll(filepath.Join(rootfs, strings.TrimPrefix(effectiveWorkdir, "/")), 0755)

	before, err := layers.SnapshotFilesystem(rootfs)
	if err != nil {
		return "", err
	}

	exitCode, err := runtime.IsolateAndRun(rootfs, []string{"/bin/sh", "-c", command}, envDict, effectiveWorkdir)
	if err != nil {
		return "", err
	}
	if exitCode != 0 {
		return "", fmt.Errorf("[BUILD ERROR] RUN failed (exit %d):\n  %s", exitCode, command)
	}

	after, err := layers.SnapshotFilesystem(rootfs)
	if err != nil {
		return "", err
	}

	changedPaths := layers.ComputeDeltaPaths(before, after, rootfs)
	tarBytes, err := layers.CreateDeltaTar(rootfs, changedPaths)
	if err != nil {
		return "", err
	}
	return layers.StoreLayer(tarBytes)
}

func copyFile(src, dst string) error {
	data, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	info, err := os.Stat(src)
	if err != nil {
		return err
	}
	return os.WriteFile(dst, data, info.Mode())
}

func copyDir(src, dst string) error {
	return filepath.Walk(src, func(p string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		rel, _ := filepath.Rel(src, p)
		target := filepath.Join(dst, rel)
		if info.IsDir() {
			return os.MkdirAll(target, info.Mode())
		}
		return copyFile(p, target)
	})
}

File: docksmith/builder.py
# docksmith/builder.py
# ============================================================
#  PROTHAM — File 3: Build Engine / Orchestrator
# ============================================================

import os
import glob
import shutil
import tempfile
import time
from datetime import datetime, timezone

from docksmith.parser import (
    parse_docksmithfile, parse_cmd_args, parse_env_args,
    parse_copy_args, parse_from_args,
)
from docksmith.layers import (
    create_delta_tar, collect_all_paths, store_layer,
    extract_layer, get_layer_size, layer_exists,
    hash_copy_sources, snapshot_filesystem, compute_delta_paths,
)


def build_image(context_dir: str, name: str, tag: str, no_cache: bool = False):
    from docksmith.state import load_manifest, save_manifest
    from docksmith.cache import CacheManager
    from docksmith.runtime import isolate_and_run
    from docksmith.models import ImageManifest, LayerEntry, ImageConfig

    docksmithfile_path = os.path.join(context_dir, "Docksmithfile")
    if not os.path.exists(docksmithfile_path):
        raise FileNotFoundError(f"[BUILD ERROR] No Docksmithfile found in: {context_dir}")

    instructions = parse_docksmithfile(docksmithfile_path)
    total_steps  = len(instructions)

    layers           = []
    env_dict         = {}
    workdir          = ""
    cmd              = []
    prev_digest      = None
    cache_busted     = False
    build_start      = datetime.now(timezone.utc)
    original_created = None

    cache = CacheManager(no_cache=no_cache)

    print()

    for step_idx, instr in enumerate(instructions, start=1):
        label = f"Step {step_idx}/{total_steps} : {instr.type} {instr.args}"

        if instr.type == "FROM":
            print(label)
            base_name, base_tag = parse_from_args(instr.args)
            base_manifest = load_manifest(base_name, base_tag)
            if base_manifest is None:
                raise RuntimeError(
                    f"[BUILD ERROR] FROM: Image '{instr.args}' not found.\n"
                    f"  Run: python3 setup_base_image.py"
                )
            layers      = list(base_manifest.layers)
            prev_digest = base_manifest.digest
            for env_str in (base_manifest.config.Env or []):
                if "=" in env_str:
                    k, _, v = env_str.partition("=")
                    env_dict[k] = v
            workdir = base_manifest.config.WorkingDir or ""
            cmd     = base_manifest.config.Cmd or []
            existing = load_manifest(name, tag)
            if existing:
                original_created = existing.created
            continue

        elif instr.type == "WORKDIR":
            print(label)
            workdir = instr.args
            continue

        elif instr.type == "ENV":
            print(label)
            k, v = parse_env_args(instr.args)
            env_dict[k] = v
            continue

        elif instr.type == "CMD":
            print(label)
            cmd = parse_cmd_args(instr.args)
            continue

        elif instr.type == "COPY":
            src_pattern      = parse_copy_args(instr.args)[0]
            dest             = parse_copy_args(instr.args)[1]
            instruction_text = f"COPY {instr.args}"
            env_serialized   = _serialize_env(env_dict)
            copy_hashes      = hash_copy_sources(src_pattern, context_dir)

            cached_digest = None
            if not cache_busted:
                cached_digest = cache.lookup(
                    prev_digest=prev_digest or "",
                    instruction_text=instruction_text,
                    workdir=workdir,
                    env_serialized=env_serialized,
                    copy_hashes=copy_hashes,
                )

            if cached_digest:
                print(f"{label} [CACHE HIT]")
                layers.append(LayerEntry(digest=cached_digest, size=get_layer_size(cached_digest), createdBy=instruction_text))
                prev_digest = cached_digest
            else:
                cache_busted = True
                t0 = time.perf_counter()
                digest = _execute_copy(src_pattern, dest, context_dir, layers, workdir)
                elapsed = time.perf_counter() - t0
                print(f"{label} [CACHE MISS] {elapsed:.2f}s")
                cache.store(prev_digest=prev_digest or "", instruction_text=instruction_text, workdir=workdir,
                            env_serialized=env_serialized, copy_hashes=copy_hashes, result_digest=digest)
                layers.append(LayerEntry(digest=digest, size=get_layer_size(digest), createdBy=instruction_text))
                prev_digest = digest

        elif instr.type == "RUN":
            instruction_text = f"RUN {instr.args}"
            env_serialized   = _serialize_env(env_dict)

            cached_digest = None
            if not cache_busted:
                cached_digest = cache.lookup(
                    prev_digest=prev_digest or "",
                    instruction_text=instruction_text,
                    workdir=workdir,
                    env_serialized=env_serialized,
                    copy_hashes=None,
                )

            if cached_digest:
                print(f"{label} [CACHE HIT]")
                layers.append(LayerEntry(digest=cached_digest, size=get_layer_size(cached_digest), createdBy=instruction_text))
                prev_digest = cached_digest
            else:
                cache_busted = True
                t0 = time.perf_counter()
                digest = _execute_run(instr.args, layers, env_dict, workdir, isolate_and_run)
                elapsed = time.perf_counter() - t0
                print(f"{label} [CACHE MISS] {elapsed:.2f}s")
                cache.store(prev_digest=prev_digest or "", instruction_text=instruction_text, workdir=workdir,
                            env_serialized=env_serialized, copy_hashes=None, result_digest=digest)
                layers.append(LayerEntry(digest=digest, size=get_layer_size(digest), createdBy=instruction_text))
                prev_digest = digest

    config = ImageConfig(
        Env=[f"{k}={v}" for k, v in env_dict.items()],
        Cmd=cmd,
        WorkingDir=workdir,
    )
    from docksmith.models import ImageConfig
    all_hits = not cache_busted
    created  = (original_created if all_hits and original_created else build_start.isoformat())

    manifest = ImageManifest(name=name, tag=tag, digest="", created=created, config=config, layers=layers)
    saved    = save_manifest(manifest)
    total_s  = (datetime.now(timezone.utc) - build_start).total_seconds()
    short    = saved.digest[:19] if saved.digest else "sha256:?"
    print(f"\nSuccessfully built {short} {name}:{tag} ({total_s:.2f}s)\n")

    # Write build log entry
    _append_build_log(name, tag, saved.digest)
    return saved


def _serialize_env(env_dict: dict) -> str:
    if not env_dict:
        return ""
    return "&".join(f"{k}={v}" for k, v in sorted(env_dict.items()))


def _assemble_rootfs(layers: list, target_dir: str):
    for layer in layers:
        extract_layer(layer.digest, target_dir)


def _execute_copy(src_pattern, dest, context_dir, current_layers, workdir) -> str:
    import glob as glob_module

    with tempfile.TemporaryDirectory(prefix="docksmith_copy_") as rootfs:
        _assemble_rootfs(current_layers, rootfs)
        if workdir:
            os.makedirs(os.path.join(rootfs, workdir.lstrip("/")), exist_ok=True)

        dest_abs = os.path.join(rootfs, dest.lstrip("/"))
        os.makedirs(dest_abs, exist_ok=True)

        if src_pattern == ".":
            matched = []
            for root, dirs, files in os.walk(context_dir):
                dirs.sort()
                for f in sorted(files):
                    matched.append(os.path.join(root, f))
        else:
            full_pattern = os.path.join(context_dir, src_pattern)
            matched = sorted(glob_module.glob(full_pattern, recursive=True))

        if not matched:
            raise FileNotFoundError(f"[COPY ERROR] No files matched '{src_pattern}' in '{context_dir}'")

        for src_path in matched:
            if os.path.isfile(src_path):
                rel     = os.path.relpath(src_path, context_dir)
                dst_dir = os.path.join(dest_abs, os.path.dirname(rel))
                os.makedirs(dst_dir, exist_ok=True)
                shutil.copy2(src_path, os.path.join(dst_dir, os.path.basename(rel)))
            elif os.path.isdir(src_path):
                shutil.copytree(src_path, os.path.join(dest_abs, os.path.basename(src_path)), dirs_exist_ok=True)

        all_paths = collect_all_paths(dest_abs)
        tar_bytes = create_delta_tar(dest_abs, all_paths)
        return store_layer(tar_bytes)


def _execute_run(command, current_layers, env_dict, workdir, isolate_fn) -> str:
    with tempfile.TemporaryDirectory(prefix="docksmith_run_") as rootfs:
        _assemble_rootfs(current_layers, rootfs)
        effective_workdir = workdir or "/"
        os.makedirs(os.path.join(rootfs, effective_workdir.lstrip("/")), exist_ok=True)

        before    = snapshot_filesystem(rootfs)
        exit_code = isolate_fn(rootfs=rootfs, command=["/bin/sh", "-c", command],
                               env=dict(env_dict), workdir=effective_workdir)
        if exit_code != 0:
            raise RuntimeError(f"[BUILD ERROR] RUN failed (exit {exit_code}):\n  {command}")

        after         = snapshot_filesystem(rootfs)
        changed_paths = compute_delta_paths(before, after, rootfs)
        tar_bytes     = create_delta_tar(rootfs, changed_paths) if changed_paths else create_delta_tar(rootfs, [])
        return store_layer(tar_bytes)


def _append_build_log(name, tag, digest):
    import json
    from datetime import datetime, timezone
    log_path = os.path.expanduser("~/.docksmith/build_log.json")
    os.makedirs(os.path.dirname(log_path), exist_ok=True)
    events = []
    if os.path.exists(log_path):
        try:
            with open(log_path) as f:
                events = json.load(f)
        except Exception:
            events = []
    events.insert(0, {
        "time":    datetime.now(timezone.utc).isoformat(),
        "image":   f"{name}:{tag}",
        "message": "built successfully",
        "status":  "built",
    })
    events = events[:50]
    with open(log_path, "w") as f:
        json.dump(events, f, indent=2)
